hsrc_lib/Database/LMDB/FFI.hsc

```diff
@@ -0,0 +1,293 @@
+        #{poke struct MDB_envinfo, me_last_pgno} p (me_last_pgno x)
+        #{poke struct MDB_envinfo, me_last_txnid} p (me_last_txnid x)
+        #{poke struct MDB_envinfo, me_maxreaders} p (me_maxreaders x)
+        #{poke struct MDB_envinfo, me_numreaders} p (me_numreaders x)
+
+-- | A value stored in the database. Be cautious; committing the
+-- transaction that obtained a value should also invalidate it;
+-- avoid capturing MDB_val in a lazy value. A safe interface
+-- similar to STRef could be provided by another module.
+data {-# CTYPE "struct MDB_val" #-} MDB_val = MDB_val {
+      mv_size :: {-# UNPACK #-} !CSize
+    , mv_data :: {-# UNPACK #-} !(Ptr ())
+    }
+
+instance Storable MDB_val where
+  sizeOf _ = #{size struct MDB_val}
+  alignment _ = #{alignment struct MDB_val}
+  peek p = do
+      mv_size <- #{peek struct MDB_val, mv_size} p
+      mv_data <- #{peek struct MDB_val, mv_data} p
+      pure $! MDB_val { mv_size, mv_data }
+  poke p x = do
+      #{poke struct MDB_val, mv_size} p (mv_size x)
+      #{poke struct MDB_val, mv_data} p (mv_data x)
+
+-- opaque types
+
+data {-# CTYPE "struct MDB_env" #-} MDB_env
+
+data {-# CTYPE "struct MDB_txn" #-} MDB_txn
+
+data {-# CTYPE "struct MDB_cursor" #-} MDB_cursor
+
+-- typedefs
+
+newtype MDB_mode_t = MDB_mode_t ( #type mdb_mode_t )
+  deriving (Show, Eq, Num, Storable)
+
+newtype MDB_dbi = MDB_dbi ( #type MDB_dbi )
+  deriving (Show, Eq, Num, Storable)
+
+{-------------------------------------------------------------------------------
+  Functions
+-------------------------------------------------------------------------------}
+
+-- FFI
+--  'safe': higher overhead, thread juggling, allows callbacks into Haskell
+--  'unsafe': lower overhead, reduced concurrency, no callbacks into Haskell
+foreign import capi unsafe "lmdb.h mdb_version" mdb_version :: Ptr CInt -> Ptr CInt -> Ptr CInt -> IO CString
+foreign import capi unsafe "lmdb.h mdb_strerror" mdb_strerror :: CInt -> CString
+
+foreign import capi "lmdb.h mdb_env_create" mdb_env_create :: Ptr (Ptr MDB_env) -> IO CInt
+foreign import capi "lmdb.h mdb_env_open" mdb_env_open :: Ptr MDB_env -> CString -> CUInt -> MDB_mode_t -> IO CInt
+foreign import capi "lmdb.h mdb_env_copy" mdb_env_copy :: Ptr MDB_env -> CString -> IO CInt
+foreign import capi "lmdb.h mdb_env_stat" mdb_env_stat :: Ptr MDB_env -> Ptr MDB_stat -> IO CInt
+foreign import capi "lmdb.h mdb_env_info" mdb_env_info :: Ptr MDB_env -> Ptr MDB_envinfo -> IO CInt
+foreign import capi "lmdb.h mdb_env_sync" mdb_env_sync :: Ptr MDB_env -> CInt -> IO CInt
+foreign import capi "lmdb.h mdb_env_close" mdb_env_close :: Ptr MDB_env -> IO ()
+foreign import capi "lmdb.h mdb_env_set_flags" mdb_env_set_flags :: Ptr MDB_env -> CUInt -> CInt -> IO CInt
+foreign import capi unsafe "lmdb.h mdb_env_get_flags" mdb_env_get_flags :: Ptr MDB_env -> Ptr CUInt -> IO CInt
+foreign import ccall unsafe "lmdb.h mdb_env_get_path" mdb_env_get_path :: Ptr MDB_env -> Ptr (Ptr CChar) -> IO CInt
```

Comment:
Perhaps use CPP to define also the `capi` version? for removing the `ccall` one eventually

Reply:
I've tried that, but then you get cascading changes, because the type of the function changes depending on the CPP branch and there are also conditional changes on imports. I think it's fine to keep this ccall for now, since we don't really this function anyway